jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/CloudWatchLogStreamCoroutine.kt

```diff
@@ -0,0 +1,133 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.cloudwatch.logs
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.project.Project
+import com.intellij.ui.table.JBTable
+import com.intellij.util.ui.ListTableModel
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.future.asDeferred
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import software.amazon.awssdk.services.cloudwatchlogs.CloudWatchLogsAsyncClient
+import software.amazon.awssdk.services.cloudwatchlogs.model.GetLogEventsRequest
+import software.amazon.awssdk.services.cloudwatchlogs.model.OutputLogEvent
+import software.aws.toolkits.jetbrains.core.awsClient
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import software.aws.toolkits.jetbrains.utils.getCoroutineUiContext
+import software.aws.toolkits.resources.message
+
+class CloudWatchLogStreamCoroutine(
+    project: Project,
+    private val table: JBTable,
+    private val logGroup: String,
+    private val logStream: String
+) : CoroutineScope by ApplicationThreadPoolScope("CloudWatchLogsStream"), Disposable {
+    val channel = Channel<Messages>()
+    private val edtContext = getCoroutineUiContext(disposable = this)
+
+    private val client: CloudWatchLogsAsyncClient = project.awsClient()
+    private var nextBackwardToken: String? = null
+    private var nextForwardToken: String? = null
+
+    enum class Messages {
+        LOAD_FORWARD,
+        LOAD_BACKWARD
+    }
+
+    suspend fun loadInitial() {
+        val request = GetLogEventsRequest.builder().logGroupName(logGroup).logStreamName(logStream).startFromHead(true).build()
+        val items = load(request, saveForwardToken = true, saveBackwardToken = true)
+        populateTable(items)
+        startListening()
+    }
+
+    suspend fun loadInitialAround(startTime: Long, duration: Long) {
+        val request = GetLogEventsRequest
+            .builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .startFromHead(true)
+            .startTime(startTime - duration)
+            .endTime(startTime + duration)
+            .build()
+        val items = load(request, saveForwardToken = true, saveBackwardToken = true)
+        populateTable(items)
+        startListening()
+    }
+
+    private suspend fun populateTable(items: List<OutputLogEvent>) {
+        if (items.isNotEmpty()) {
+            val events = table.logsModel.items.plus(items)
+            withContext(edtContext) { table.logsModel.items = events }
+        }
+        table.emptyText.text = message("cloudwatch.logs.no_events")
```

Comment:
Do we need to set this everytime? Can we recover from an error (where we change this to the error text) or is the error a terminal state?

Reply:
Good point, we could show a refresh button to recover on failure. I'd like to put that in a subsequent PR instead of this one though